### PR TITLE
feat(SpokePool): Specify exclusivityDeadlineOffset by default

### PR DESCRIPTION
**Pull request by nicholaspai**
Better depositor UX if they can specify an offset that is relative to the time that their deposit mines, so they do not have to account for any off-chain latency

### CODE REVIEW

File: contracts/SpokePool.sol [hunk 1]
```diff
@@ -569,15 +570,15 @@ abstract contract SpokePool is
 
         // As a safety measure, prevent caller from inadvertently locking funds during exclusivity period
         //  by forcing them to specify an exclusive relayer if the exclusivity period
```
Comment (pxrl) [inline contracts/SpokePool.sol:572]: Driveby observation.

```suggestion
        // by forcing them to specify an exclusive relayer if the exclusivity period
```

File: contracts/SpokePool.sol [hunk 2]
```diff
@@ -569,15 +570,15 @@ abstract contract SpokePool is
...
-        // is in the future. If this deadline is 0, then the exclusive relayer must also be address(0).
+        // is in the future. If this offset is 0, then the exclusive relayer must also be address(0).
         // @dev Checks if either are > 0 by bitwise or-ing.
-        if (uint256(uint160(exclusiveRelayer)) | exclusivityDeadline != 0) {
+        if (uint256(uint160(exclusiveRelayer)) | exclusivityDeadlineOffset != 0) {
             // Now that we know one is nonzero, we need to perform checks on each.
             // Check that exclusive relayer is nonzero.
             if (exclusiveRelayer == address(0)) revert InvalidExclusiveRelayer();
```
Comment (pxrl) [inline contracts/SpokePool.sol:578]: ooc, do we need to retain this check? It prevents a depositor from setting 0x0 as the exclusiveRelayer, but it doesn't prevent the rest of the invalid (i.e. non-relayer) address space from being used. I think the main function would be to catch depositors accidentally setting this field, but that seems like an error that an integrator would quickly identify and fix because none of their testing would succeed.

```suggestion
            // Now that we know one is nonzero, we need to perform checks on each.
            // Check that exclusive relayer is nonzero.
            if (exclusiveRelayer == address(0)) revert InvalidExclusiveRelayer();
```

File: contracts/SpokePool.sol [hunk 3]
```diff
@@ -540,7 +541,7 @@ abstract contract SpokePool is
         address exclusiveRelayer,
         uint32 quoteTimestamp,
         uint32 fillDeadline,
-        uint32 exclusivityDeadline,
+        uint32 exclusivityDeadlineOffset,
```
Comment (pxrl) [inline contracts/SpokePool.sol:544]: From the perspective of a depositor I think `exclusivityWindow` or `exclusivityPeriod` might be more self-explanatory than `exclusivityDeadlineOffset`.

File: contracts/SpokePool.sol [hunk 4]
```diff
@@ -642,9 +643,9 @@ abstract contract SpokePool is
      * @param fillDeadlineOffset Added to the current time to set the fill deadline, which is the deadline for the
      * relayer to fill the deposit. After this destination chain timestamp, the fill will revert on the
      * destination chain.
-     * @param exclusivityDeadline The latest timestamp that only the exclusive relayer can fill the deposit. After this
-     * destination chain timestamp, anyone can fill this deposit on the destination chain. Should be 0 if
-     * exclusive relayer is 0.
+     * @param exclusivityDeadlineOffset Added to the current time to set the exclusive relayer deadline,
+     * which is the deadline for the exclusiveRelayer to fill the deposit. After this destination chain timestamp,
+     * anyone can fill the deposit.
```
Comment (pxrl) [inline contracts/SpokePool.sol:648]:
```suggestion
     * anyone can fill the deposit up to the fillDeadline timestamp.
```

File: contracts/SpokePool.sol [hunk 5]
```diff
@@ -672,73 +667,6 @@ abstract contract SpokePool is
             exclusiveRelayer,
             uint32(getCurrentTime()),
             uint32(getCurrentTime()) + fillDeadlineOffset,
-            exclusivityDeadline,
```
Comment (mrice32) [inline contracts/SpokePool.sol:675]: nit: can we add this back? Just to reduce the diff size? I think we have the bytecode space regardless, no?

File: contracts/SpokePool.sol (hunk 3 above)
Comment (mrice32) [reply contracts/SpokePool.sol:544]: I think period is the best, yeah

File: contracts/SpokePool.sol (hunk 2 above)
Comment (mrice32) [reply contracts/SpokePool.sol]: Maybe not.

Nick and I spoke on this today, and we thought the only check that we may want to leave in is what's there now. Basically to not allow a user to set the exclusive relayer to 0 along with a nonzero exclusivity deadline, since that's clearly wrong. But we could also drop that one too just for simplicity, especially if we use 0x0 exclusive relayer as the on/off switch in the fill function.

Comment (nicholaspai) [reply contracts/SpokePool.sol]: what do you think of the current check now @pxrl ? I'm not sure what your original push for was but the diff has updated

Comment (pxrl) [reply contracts/SpokePool.sol]: This is simpler than what was there before :+1: I do still wonder if we could drop the check entirely, since the probability of funds locking funds is so low and so temporary.

File: contracts/SpokePool.sol [hunk 6]
```diff
@@ -568,16 +569,10 @@ abstract contract SpokePool is
...
-        // is in the future. If this deadline is 0, then the exclusive relayer must also be address(0).
-        // @dev Checks if either are > 0 by bitwise or-ing.
-        if (uint256(uint160(exclusiveRelayer)) | exclusivityDeadline != 0) {
-            // Now that we know one is nonzero, we need to perform checks on each.
-            // Check that exclusive relayer is nonzero.
-            if (exclusiveRelayer == address(0)) revert InvalidExclusiveRelayer();
-
-            // Check that deadline is in the future.
-            if (exclusivityDeadline < currentTime) revert InvalidExclusivityDeadline();
+        //  by forcing them to specify an exclusive relayer if the exclusivity deadline timestamp
```
Comment (pxrl) [inline contracts/SpokePool.sol:572]:
```suggestion
        // by forcing them to specify an exclusive relayer if the exclusivity deadline timestamp
```

**Review by pxrl — APPROVED** [review]

File: contracts/SpokePool.sol (hunk 2 above)
Comment (mrice32) [reply contracts/SpokePool.sol]: Yeah, the only thing this check protects against now is someone accidentally setting an exclusivity timestamp, but not a relayer. It's sort of just a bug signal since there is no reason anyone would want to do that.

I think removing is probably better -- saves gas, and this check doesn't really protect against anything.

**Review by mrice32 — APPROVED** [review]

File: contracts/SpokePool.sol [hunk 7]
```diff
@@ -845,7 +841,11 @@ abstract contract SpokePool is
     {
         // Exclusivity deadline is inclusive and is the latest timestamp that the exclusive relayer has sole right
         // to fill the relay.
-        if (relayData.exclusivityDeadline >= getCurrentTime() && relayData.exclusiveRelayer != msg.sender) {
+        if (
+            relayData.exclusiveRelayer != address(0) &&
+            relayData.exclusivityDeadline >= getCurrentTime() &&
+            relayData.exclusiveRelayer != msg.sender
```
Comment (mrice32) [inline contracts/SpokePool.sol:847]: nit: can we move this to the first check. Small gas optimization.

We think exclusivity will be the most common case, so we probably expect that most fills will be able to exit this if on the `relayData.exclusiveRelayer != msg.sender` check.

Comment (mrice32) [inline contracts/SpokePool.sol:836]:
```suggestion
            relayData.exclusiveRelayer != msg.sender &&
            relayData.exclusivityDeadline >= getCurrentTime() &&
            relayData.exclusiveRelayer != address(0)
```
